File: android/src/main/kotlin/dev/doclens/doclens/DoclensPlugin.kt
package dev.doclens.doclens

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import io.flutter.view.TextureRegistry
import java.util.concurrent.Executors

private const val CAMERA_PERMISSION_REQUEST = 0x4F4E

class DoclensPlugin :
    FlutterPlugin,
    ActivityAware,
    MethodChannel.MethodCallHandler,
    PluginRegistry.RequestPermissionsResultListener {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var textureRegistry: TextureRegistry
    private lateinit var context: Context

    private val eventStream = QuadEventStream()
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val backgroundExecutor = Executors.newSingleThreadExecutor()

    private var session: ScannerSession? = null
    private var pendingPermissionInit: (() -> Unit)? = null
    private var nativeUIScanner: NativeUIScanner? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        textureRegistry = binding.textureRegistry
        methodChannel = MethodChannel(binding.binaryMessenger, "doclens/methods")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "doclens/events")
        eventChannel.setStreamHandler(eventStream)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        session?.dispose()
        session = null
        backgroundExecutor.shutdown()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addRequestPermissionsResultListener(this)
        val ui = NativeUIScanner(context).also { nativeUIScanner = it }
        binding.addActivityResultListener(ui)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
        onAttachedToActivity(binding)

    override fun onDetachedFromActivity() {
        activityBinding?.removeRequestPermissionsResultListener(this)
        nativeUIScanner?.let { activityBinding?.removeActivityResultListener(it) }
        nativeUIScanner = null
        activity = null
        activityBinding = null
    }

    override fun onDetachedFromActivityForConfigChanges() = onDetachedFromActivity()

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> initialize(call, result)
            "dispose" -> { session?.dispose(); session = null; result.success(null) }
            "capture" -> {
                val s = session
                if (s == null) { result.error("init_failed", "Not initialized", null); return }
                s.capture(result)
            }
            "warpImage" -> handleWarp(call, result)
            "setFlashMode" -> { session?.setFlashMode(call.argument<String>("mode") ?: "auto"); result.success(null) }
            "switchCamera" -> { session?.switchCamera(); result.success(null) }
            "pause" -> { session?.pause(); result.success(null) }
            "resume" -> { session?.resume(); result.success(null) }
            "focusAt" -> {
                val x = (call.argument<Number>("x") ?: 0.5).toFloat()
                val y = (call.argument<Number>("y") ?: 0.5).toFloat()
                session?.focus(x, y)
                result.success(null)
            }
            "scanWithNativeUI" -> handleNativeUIScan(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleNativeUIScan(call: MethodCall, result: MethodChannel.Result) {
        val activity = this.activity
        val ui = nativeUIScanner
        if (activity == null || ui == null) {
            result.error("init_failed", "No activity attached", null); return
        }
        ui.launch(
            activity = activity,
            pageLimit = (call.argument<Int>("pageLimit") ?: 100),
            allowGalleryImport = call.argument<Boolean>("allowGalleryImport") ?: false,
            jpegQuality = call.argument<Int>("jpegQuality") ?: 100,
            result = result,
        )
    }

    private fun handleWarp(call: MethodCall, result: MethodChannel.Result) {
        val rawPath = call.argument<String>("rawImagePath")
        val quadMap = call.argument<Map<String, Any?>>("quad")
        val quality = call.argument<Int>("jpegQuality") ?: 100
        if (rawPath == null || quadMap == null) {
            result.error("capture_failed", "Invalid warp args", null); return
        }
        backgroundExecutor.execute {
            try {
                val q = Quad.fromMap(quadMap)
                val out = ImageWarper.warpFile(rawPath, q, quality)
                mainHandler.post { result.success(out) }
            } catch (e: Exception) {
                mainHandler.post { result.error("capture_failed", e.message, null) }
            }
        }
    }

    private fun initialize(call: MethodCall, result: MethodChannel.Result) {
        val activity = this.activity
        if (activity == null) {
            result.error("init_failed", "No activity attached", null); return
        }
        // Tear down any previous session — never leak a running camera.
        session?.dispose()
        session = null

        val granted = ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) ==
            PackageManager.PERMISSION_GRANTED
        if (!granted) {
            pendingPermissionInit = { doInitialize(call, result) }
            ActivityCompat.requestPermissions(
                activity,
                arrayOf(Manifest.permission.CAMERA),
                CAMERA_PERMISSION_REQUEST,
            )
            return
        }
        doInitialize(call, result)
    }

    @Suppress("UNCHECKED_CAST")
    private fun doInitialize(call: MethodCall, result: MethodChannel.Result) {
        val activity = this.activity ?: run {
            result.error("init_failed", "No activity attached", null); return
        }
        val cfg = ScannerConfig.fromMap(call.arguments as? Map<String, Any?> ?: emptyMap())
        val entry = textureRegistry.createSurfaceTexture()
        val s = ScannerSession(
            activity = activity,
            context = context,
            config = cfg,
            textureEntry = entry,
            eventSink = { eventStream.send(it) },
        )
        session = s
        s.start { startResult ->
            startResult.fold(
                onSuccess = { result.success(it) },
                onFailure = { err ->
                    val code = when (err) {
                        is ScannerException.PermissionDenied -> "permission_denied"
                        is ScannerException.Unavailable -> "unavailable"
                        else -> "init_failed"
                    }
                    entry.release()
                    session = null
                    result.error(code, err.message, null)
                },
            )
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
    ): Boolean {
        if (requestCode != CAMERA_PERMISSION_REQUEST) return false
        val pending = pendingPermissionInit ?: return true
        pendingPermissionInit = null
        // Either way, invoke the pending init: on grant it will start the
        // session normally; on denial `ScannerSession.start` re-checks
        // permission and surfaces `permission_denied` via the Result's
        // failure path.
        mainHandler.post(pending)
        return true
    }
}

class QuadEventStream : EventChannel.StreamHandler {
    @Volatile private var sink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) { sink = events }
    override fun onCancel(arguments: Any?) { sink = null }

    fun send(payload: Map<String, Any?>) {
        val s = sink ?: return
        mainHandler.post {
            try { s.success(payload) } catch (_: Exception) {}
        }
    }
}

File: ios/doclens/Sources/doclens/DoclensPlugin.swift
import Flutter
import UIKit

public class DoclensPlugin: NSObject, FlutterPlugin {
    private var session: ScannerSession?
    private let registrar: FlutterPluginRegistrar
    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    private let eventStream = QuadEventStream()
    private let nativeUIScanner = NativeUIScanner()

    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        self.methodChannel = FlutterMethodChannel(
            name: "doclens/methods",
            binaryMessenger: registrar.messenger())
        self.eventChannel = FlutterEventChannel(
            name: "doclens/events",
            binaryMessenger: registrar.messenger())
        super.init()
        registrar.addMethodCallDelegate(self, channel: methodChannel)
        eventChannel.setStreamHandler(eventStream)
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let plugin = DoclensPlugin(registrar: registrar)
        registrar.publish(plugin)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            initialize(args: call.arguments as? [String: Any] ?? [:], result: result)
        case "dispose":
            session?.stop()
            session = nil
            eventStream.reset()
            result(nil)
        case "capture":
            guard let session = session else {
                result(FlutterError(code: "init_failed", message: "Not initialized", details: nil))
                return
            }
            session.capture { res in
                switch res {
                case .success(let payload): result(payload)
                case .failure(let err):
                    result(FlutterError(code: "capture_failed",
                                        message: err.localizedDescription,
                                        details: nil))
                }
            }
        case "warpImage":
            handleWarp(call: call, result: result)
        case "rotateImage":
            handleRotate(call: call, result: result)
        case "recognizeText":
            handleRecognizeText(call: call, result: result)
        case "setFlashMode":
            if let mode = (call.arguments as? [String: Any])?["mode"] as? String {
                session?.setFlashMode(mode)
            }
            result(nil)
        case "switchCamera":
            session?.switchCamera()
            result(nil)
        case "pause":
            session?.pause()
            result(nil)
        case "resume":
            session?.resume()
            result(nil)
        case "focusAt":
            let args = call.arguments as? [String: Any] ?? [:]
            let x = (args["x"] as? Double) ?? 0.5
            let y = (args["y"] as? Double) ?? 0.5
            session?.focus(at: CGPoint(x: x, y: y))
            result(nil)
        case "scanWithNativeUI":
            let args = call.arguments as? [String: Any] ?? [:]
            let quality = (args["jpegQuality"] as? Int) ?? 100
            nativeUIScanner.present(jpegQuality: quality, result: result)
        case "detectInImage":
            handleDetectInImage(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize(args: [String: Any], result: @escaping FlutterResult) {
        // Tear down any previous session so we never leak a running camera.
        session?.stop()
        session = nil
        let config = ScannerConfig.from(map: args)
        let textures = registrar.textures()
        let new = ScannerSession(config: config, textures: textures,
                                 eventSink: { [weak self] payload in
            self?.eventStream.send(payload)
        })
        new.start { startResult in
            switch startResult {
            case .success(let textureId):
                self.session = new
                result(NSNumber(value: textureId))
            case .failure(let err):
                let code: String
                if case .permissionDenied = err { code = "permission_denied" }
                else if case .unavailable = err { code = "unavailable" }
                else { code = "init_failed" }
                result(FlutterError(code: code,
                                    message: err.localizedDescription,
                                    details: nil))
            }
        }
    }

    private func handleWarp(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["rawImagePath"] as? String,
              let quadMap = args["quad"] as? [String: Any] else {
            result(FlutterError(code: "capture_failed",
                                message: "Invalid warp args", details: nil))
            return
        }
        let jpegQuality = (args["jpegQuality"] as? Int) ?? 100
        let enhancement = (args["enhancement"] as? String) ?? "none"
        let autoOrientation = (args["autoOrientation"] as? String) ?? "none"
        let quad = Quad.fromMap(quadMap)
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let out = try ImageWarper.warpFile(
                    inputPath: path, quad: quad, jpegQuality: jpegQuality,
                    enhancement: enhancement, autoOrientation: autoOrientation)
                DispatchQueue.main.async { result(out) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "capture_failed",
                                        message: error.localizedDescription,
                                        details: nil))
                }
            }
        }
    }

    private func handleRecognizeText(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["imagePath"] as? String else {
            result(FlutterError(code: "capture_failed",
                                message: "Invalid recognizeText args", details: nil))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let out = try TextRecognizer.recognize(path: path)
                DispatchQueue.main.async { result(out) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "capture_failed",
                                        message: error.localizedDescription,
                                        details: nil))
                }
            }
        }
    }

    private func handleDetectInImage(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["imagePath"] as? String else {
            result(FlutterError(code: "capture_failed",
                                message: "Invalid detectInImage args", details: nil))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            // Bake EXIF orientation into the pixels so the detected quad and
            // reported size live in the same upright space the warp uses.
            guard let img = UIImage(contentsOfFile: path),
                  let cg = img.uprightCGImage() else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "capture_failed",
                                        message: "Cannot read image at \(path)",
                                        details: nil))
                }
                return
            }
            let size = CGSize(width: cg.width, height: cg.height)
            DocumentDetector.detect(cgImage: cg) { quad in
                let payload: [String: Any] = [
                    "quad": quad?.toMap() ?? NSNull(),
                    "imageSize": [Double(size.width), Double(size.height)],
                ]
                DispatchQueue.main.async { result(payload) }
            }
        }
    }

    private func handleRotate(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["imagePath"] as? String else {
            result(FlutterError(code: "capture_failed",
                                message: "Invalid rotate args", details: nil))
            return
        }
        let quarterTurns = (args["quarterTurns"] as? Int) ?? 0
        let jpegQuality = (args["jpegQuality"] as? Int) ?? 100
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let out = try ImageWarper.rotateFile(
                    inputPath: path, quarterTurns: quarterTurns, jpegQuality: jpegQuality)
                DispatchQueue.main.async { result(out) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "capture_failed",
                                        message: error.localizedDescription,
                                        details: nil))
                }
            }
        }
    }
}

final class QuadEventStream: NSObject, FlutterStreamHandler {
    private var sink: FlutterEventSink?

    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }

    func send(_ payload: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.sink?(payload)
        }
    }

    func reset() {
        DispatchQueue.main.async { [weak self] in
            self?.sink = nil
        }
    }
}
